components/ai_chat/browser/ai_chat_database.cc

```diff
@@ -0,0 +1,271 @@
+  sql::Statement statement(GetDB().GetUniqueStatement(
+      "SELECT conversation.*, entries.date "
+      "FROM conversation"
+      " JOIN ("
+      "SELECT conversation_id, date"
+      " FROM conversation_entry"
+      " GROUP BY conversation_id"
+      " ORDER BY date DESC"
+      ") AS entries"
+      " ON conversation.id = entries.conversation_id"));
+
+  std::vector<mojom::ConversationPtr> conversation_list;
+
+  while (statement.Step()) {
+    mojom::Conversation conversation;
+    conversation.id = statement.ColumnInt64(0);
+    conversation.title = statement.ColumnString(1);
+    conversation.page_url = GURL(statement.ColumnString(2));
+    conversation.date = DeserializeTime(statement.ColumnInt64(3));
+    conversation_list.emplace_back(conversation.Clone());
+  }
+
+  return conversation_list;
+}
+
+std::vector<mojom::ConversationEntryPtr> AIChatDatabase::GetConversationEntry(
+    int64_t conversation_id) {
+  sql::Statement statement(GetDB().GetUniqueStatement(
+      "SELECT conversation_entry.*, entry_text.* "
+      "FROM conversation_entry"
+      " JOIN ("
+      " SELECT *"
+      " FROM conversation_entry_text"
+      ") AS entry_text"
+      " ON conversation_entry.id = entry_text.conversation_entry_id"
+      " WHERE conversation_id=?"
+      " ORDER BY conversation_entry.date DESC"));
+
+  statement.BindInt64(0, conversation_id);
+
+  std::vector<mojom::ConversationEntryPtr> history;
+
+  while (statement.Step()) {
+    mojom::ConversationEntryText entry_text;
+    entry_text.id = statement.ColumnInt64(4);
+    entry_text.text = statement.ColumnString(5);
+    int64_t conversation_entry_id = statement.ColumnInt64(6);
+
+    auto iter = base::ranges::find_if(
+        history,
+        [&conversation_entry_id](const mojom::ConversationEntryPtr& entry) {
+          return entry->id = conversation_entry_id;
+        });
+
+    // A ConversationEntry can include multiple generated texts
+    if (iter != history.end()) {
+      (*iter)->texts.emplace_back(entry_text.Clone());
+      continue;
+    }
+
+    mojom::ConversationEntry entry;
+    entry.id = statement.ColumnInt64(0);
+    entry.date = DeserializeTime(statement.ColumnInt64(1));
+    entry.character_type =
+        static_cast<mojom::CharacterType>(statement.ColumnInt(2));
+    entry.texts.emplace_back(entry_text.Clone());
+
+    history.emplace_back(entry.Clone());
+  }
+
+  return history;
+}
+
+bool AIChatDatabase::AddConversation(mojom::ConversationPtr conversation) {
+  sql::Statement statement(GetDB().GetUniqueStatement(
+      "INSERT INTO conversation(" CONVERSATION_ROW_FIELDS
+      ") VALUES(NULL, ?, ?)"));
+
+  statement.BindString(0, conversation->title);
+  if (conversation->page_url.has_value()) {
+    statement.BindString(1, conversation->page_url->spec());
+  }
+
+  if (!statement.Run()) {
+    DVLOG(0) << "Failed to execute 'conversation' insert statement";
+    return false;
+  }
+
+  return true;
```

Comment:
This is now returning the row id. 

I could use `GetLastInsertRowId()` however this could be unsafe if multiple threads are writing at the same time. There's a feature called `RETURNING` clause which I'm using: https://www3.sqlite.org/lang_returning.html